File: stream-chat-android-docs/src/main/java/io/getstream/chat/docs/java/ui/channels/ChannelList.java
package io.getstream.chat.docs.java.ui.channels;

import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import java.util.Collections;

import io.getstream.chat.android.client.ChatClient;
import io.getstream.chat.android.models.Channel;
import io.getstream.chat.android.models.FilterObject;
import io.getstream.chat.android.models.Filters;
import io.getstream.chat.android.models.querysort.QuerySortByField;
import io.getstream.chat.android.ui.ChatUI;
import io.getstream.chat.android.ui.feature.channels.list.ChannelListView;
import io.getstream.chat.android.ui.feature.channels.list.adapter.ChannelListItem;
import io.getstream.chat.android.ui.feature.channels.list.adapter.ChannelListPayloadDiff;
import io.getstream.chat.android.ui.feature.channels.list.adapter.viewholder.BaseChannelListItemViewHolder;
import io.getstream.chat.android.ui.feature.channels.list.adapter.viewholder.ChannelListItemViewHolderFactory;
import io.getstream.chat.android.ui.helper.TransformStyle;
import io.getstream.chat.android.ui.viewmodel.channels.ChannelListViewModel;
import io.getstream.chat.android.ui.viewmodel.channels.ChannelListViewModelBinding;
import io.getstream.chat.android.ui.viewmodel.channels.ChannelListViewModelFactory;
import io.getstream.chat.docs.R;
import io.getstream.chat.docs.databinding.CustomChannelListItemBinding;

/**
 * [Channel List](https://getstream.io/chat/docs/sdk/android/ui/channel-components/channel-list/)
 */
public class ChannelList extends Fragment {

    private ChannelListView channelListView;

    /**
     * [Usage](https://getstream.io/chat/docs/sdk/android/ui/channel-components/channel-list/#usage)
     */
    public void usage() {
        // Instantiate the ViewModel
        FilterObject filter = Filters.and(
                Filters.eq("type", "messaging"),
                Filters.in("members", Collections.singletonList(ChatClient.instance().getCurrentUser().getId()))
        );

        ViewModelProvider.Factory factory = new ChannelListViewModelFactory.Builder()
                .filter(filter)
                .sort(QuerySortByField.descByName("last_updated"))
                .limit(30)
                .build();

        ChannelListViewModel viewModel = new ViewModelProvider(this, factory).get(ChannelListViewModel.class);

        // Bind the ViewModel with ChannelListView
        ChannelListViewModelBinding.bind(viewModel, channelListView, getViewLifecycleOwner());
    }

    /**
     * [Handling Actions](https://getstream.io/chat/docs/sdk/android/ui/channel-components/channel-list/#handling-actions)
     */
    public void handlingActions() {
        channelListView.setChannelItemClickListener(channel -> {
            // Handle channel click
        });
        channelListView.setChannelInfoClickListener(channel -> {
            // Handle channel info click
        });
        channelListView.setUserClickListener(user -> {
            // Handle member click
        });
    }

    /**
     * [Customization](https://getstream.io/chat/docs/sdk/android/ui/channel-components/channel-list/#customization)
     */
    public void customization() {
        TransformStyle.setChannelListStyleTransformer(source -> {
            // Customize the style
            return source;
        });
    }

    /**
     * [Creating a Custom ViewHolder Factory](https://getstream.io/chat/docs/sdk/android/ui/channel-components/channel-list/#creating-a-custom-viewholder-factory)
     */
    class CustomViewHolderFactory {

        public class CustomChannelListItemViewHolderFactory extends ChannelListItemViewHolderFactory {
            @NonNull
            @Override
            protected BaseChannelListItemViewHolder createChannelViewHolder(@NonNull ViewGroup parent) {
                CustomChannelListItemBinding binding = CustomChannelListItemBinding
                        .inflate(LayoutInflater.from(parent.getContext()), parent, false);
                return new CustomChannelViewHolder(binding, getListenerContainer().getChannelClickListener());
            }
        }

        public class CustomChannelViewHolder extends BaseChannelListItemViewHolder {

            private CustomChannelListItemBinding binding;

            private Channel channel;

            public CustomChannelViewHolder(CustomChannelListItemBinding binding,
                                           ChannelListView.ChannelClickListener channelClickListener) {
                super(binding.getRoot());
                this.binding = binding;

                binding.getRoot().setOnClickListener(v -> channelClickListener.onClick(channel));
            }

            @Override
            public void bind(@NonNull ChannelListItem.ChannelItem channelItem, @NonNull ChannelListPayloadDiff diff) {
                this.channel = channelItem.getChannel();
                binding.channelAvatarView.setChannel(channel);
                String channelName = ChatUI.getChannelNameFormatter().formatChannelName(
                        channel,
                        ChatClient.instance().getCurrentUser()
                );
                binding.channelNameTextView.setText(channelName);
                String memberCount = itemView.getContext().getResources().getQuantityString(
                        R.plurals.members_count,
                        channel.getMembers().size(),
                        channel.getMembers().size()
                );
                binding.membersCountTextView.setText(memberCount);
            }
        }

        public void settingCustomViewHolderFactory() {
            // Create custom view holder factory
            CustomChannelListItemViewHolderFactory customFactory = new CustomChannelListItemViewHolderFactory();

            // Set custom view holder factory
            channelListView.setViewHolderFactory(customFactory);
        }
    }

    /**
     * [Creating a Custom Loading View](https://getstream.io/chat/docs/sdk/android/ui/channel-components/channel-list/#creating-a-custom-loading-view)
     */
    public void customizingLoadingView() {
        // Inflate loading view
        View loadingView = LayoutInflater.from(getContext()).inflate(R.layout.channel_list_loading_view, null);
        // Set loading view
        channelListView.setLoadingView(loadingView, new FrameLayout.LayoutParams(MATCH_PARENT, MATCH_PARENT));
    }
}

File: stream-chat-android-docs/src/main/kotlin/io/getstream/chat/docs/kotlin/ui/channels/ChannelList.kt
package io.getstream.chat.docs.kotlin.ui.channels

import android.content.Context
import android.graphics.Color
import android.view.LayoutInflater
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import android.widget.FrameLayout
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import io.getstream.chat.android.client.ChatClient
import io.getstream.chat.android.models.Channel
import io.getstream.chat.android.models.Filters
import io.getstream.chat.android.models.querysort.QuerySortByField
import io.getstream.chat.android.ui.ChatUI
import io.getstream.chat.android.ui.feature.channels.list.ChannelListView
import io.getstream.chat.android.ui.feature.channels.list.adapter.ChannelListPayloadDiff
import io.getstream.chat.android.ui.feature.channels.list.adapter.viewholder.BaseChannelListItemViewHolder
import io.getstream.chat.android.ui.feature.channels.list.adapter.viewholder.ChannelListItemViewHolderFactory
import io.getstream.chat.android.ui.helper.StyleTransformer
import io.getstream.chat.android.ui.helper.TransformStyle
import io.getstream.chat.android.ui.viewmodel.channels.ChannelListViewModel
import io.getstream.chat.android.ui.viewmodel.channels.ChannelListViewModelFactory
import io.getstream.chat.android.ui.viewmodel.channels.bindView
import io.getstream.chat.docs.R
import io.getstream.chat.docs.databinding.CustomChannelListItemBinding

/**
 * [Channel List](https://getstream.io/chat/docs/sdk/android/ui/channel-components/channel-list/)
 */
private class ChannelList : Fragment() {

    private lateinit var channelListView: ChannelListView

    /**
     * [Usage](https://getstream.io/chat/docs/sdk/android/ui/channel-components/channel-list/#usage)
     */
    fun usage() {
        // Instantiate the ViewModel
        val viewModel: ChannelListViewModel by viewModels {
            ChannelListViewModelFactory(
                filter = Filters.and(
                    Filters.eq("type", "messaging"),
                    Filters.`in`("members", listOf(ChatClient.instance().getCurrentUser()!!.id)),
                ),
                sort = QuerySortByField.descByName("last_updated"),
                limit = 30,
            )
        }
        // Bind the ViewModel with ChannelListView
        viewModel.bindView(channelListView, viewLifecycleOwner)
    }

    /**
     * [Handling Actions](https://getstream.io/chat/docs/sdk/android/ui/channel-components/channel-list/#handling-actions)
     */
    fun handlingActions() {
        channelListView.setChannelItemClickListener { channel ->
            // Handle channel click
        }
        channelListView.setChannelInfoClickListener { channel ->
            // Handle channel info click
        }
        channelListView.setUserClickListener { user ->
            // Handle member click
        }
    }

    /**
     * [Customization](https://getstream.io/chat/docs/sdk/android/ui/channel-components/channel-list/#customization)
     */
    fun customization(context: Context) {
        TransformStyle.channelListStyleTransformer = StyleTransformer { defaultStyle ->
            defaultStyle.copy(
                optionsEnabled = false,
                foregroundLayoutColor = Color.LTGRAY,
                indicatorReadIcon = ContextCompat.getDrawable(context, R.drawable.stream_ui_ic_flag)!!,
                channelTitleText = defaultStyle.channelTitleText.copy(
                    color = Color.BLUE,
                    size = context.resources.getDimensionPixelSize(R.dimen.stream_ui_text_large),
                ),
                unreadMessageCounterBackgroundColor = Color.BLUE,
            )
        }
    }

    /**
     * [Creating a Custom ViewHolder Factory](https://getstream.io/chat/docs/sdk/android/ui/channel-components/channel-list/#creating-a-custom-viewholder-factory)
     */
    class CustomViewHolderFactory {

        class CustomChannelListItemViewHolderFactory : ChannelListItemViewHolderFactory() {
            override fun createChannelViewHolder(parentView: ViewGroup): BaseChannelListItemViewHolder {
                return CustomChannelViewHolder(parentView, listenerContainer.channelClickListener)
            }
        }

        class CustomChannelViewHolder(
            parent: ViewGroup,
            private val channelClickListener: ChannelListView.ChannelClickListener,
            private val binding: CustomChannelListItemBinding = CustomChannelListItemBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            ),
        ) : BaseChannelListItemViewHolder(binding.root) {

            private lateinit var channel: Channel

            init {
                binding.root.setOnClickListener { channelClickListener.onClick(channel) }
            }

            override fun bind(channel: Channel, diff: ChannelListPayloadDiff) {
                this.channel = channel

                binding.channelAvatarView.setChannel(channel)
                binding.channelNameTextView.text = ChatUI.channelNameFormatter.formatChannelName(
                    channel = channel,
                    currentUser = ChatClient.instance().getCurrentUser()
                )
                binding.membersCountTextView.text = itemView.context.resources.getQuantityString(
                    R.plurals.members_count,
                    channel.members.size,
                    channel.members.size
                )
            }
        }

        fun settingCustomViewHolderFactory(channelListView: ChannelListView) {
            // Create custom view holder factory
            val customFactory = CustomChannelListItemViewHolderFactory()

            // Set custom view holder factory
            channelListView.setViewHolderFactory(customFactory)
        }
    }

    /**
     * [Creating a Custom Loading View](https://getstream.io/chat/docs/sdk/android/ui/channel-components/channel-list/#creating-a-custom-loading-view)
     */
    fun customizingLoadingView() {
        // Inflate loading view
        val loadingView = LayoutInflater.from(context).inflate(R.layout.channel_list_loading_view, null)
        // Set loading view
        channelListView.setLoadingView(loadingView, FrameLayout.LayoutParams(MATCH_PARENT, MATCH_PARENT))
    }
}
